Separable module configurations

becker33 opened the pull request:
Currently, module configurations are inconsistent because modulefiles are generated with the configs for the active environment, but are shared among all environments (and spack outside any environment).

This PR fixes that by allowing Spack environments (or other spack config scopes) to define additional sets of modules to generate. Each set of modules can enable either lmod or tcl modules, and contains all of the previously available module configuration. The user defines the name of each module set -- the set configured in Spack by default is named "default", and is the one returned by module manipulation commands in the absence of user intervention.

As part of this change, the module roots configuration moved from the `config` section to inside each module configuration.

Additionally, it adds a feature that the modulefiles for an environment can be configured to be relative to an environment view rather than the underlying prefix. This will not be enabled by default, as it should only be enabled within an environment and for non-default views constructed with separate projections per-spec.

TODO:
- [x] code changes to support multiple module sets
- [x] code changes to support modules relative to a view
- [x] Tests for multiple module configurations
- [x] Tests for modules relative to a view
- [x] Backwards compatibility for module roots from config section
- [x] Backwards compatibility for default module set without the name specified
- [x] Tests for backwards compatibility
- [ ] Documentation changes

becker33 committed:
allow multiple defined sets of module configs

becker33 committed:
backwards compatibility for module roots

becker33 committed:
update bash completions

tgamblin commented:
I think this:

> Spack environments (or other spack config sections)

should say "or other spack config *scopes*", right?

becker33 commented:
@tgamblin yes it should. Fixed.

tgamblin commented on etc/spack/defaults/modules.yaml:
```diff
@@ -14,26 +14,35 @@
-    '':
-      - CMAKE_PREFIX_PATH
+  # These are configurations for the module set named "default"
+  default:
```
I can't decide whether this should be called `default`, as yes it is default but we will override it.  It's weird to talk about "non-default changes to the default configuration" -- or at least I think that there are many ways this can be overloaded in conversation.  If we say "builtin", I guess that is also bad, and it is overloaded with the `builtin` repo.  I can't think of a good name for it, really -- is `default` the best we can do?

tgamblin replied:
FWIW, I think I might've gotten used to `default` since I wrote this.

alalazo replied:
This is definitely an improvement over the current state, so :heavy_check_mark: from my side. 

I was wondering though if we considered isolating the configuration even further by allowing a list of entries, each of which generates just a single type of modules:
```yaml
modules:
- name: 'tcl-default'
  type: tcl
  root: $spack/share/spack/modules
  [...]
- name: 'lmod-default'
  type: lmod
  root: $spack/share/spack/lmod
  [...]
```
This will simplify handling roots and remove the necessity to use `enable`.

alalazo replied:
I guess a drawback of my suggestion is that having different defaults based on platforms would require more extensive changes to the current code :thinking: Anyhow, left the idea here if that is considered useful.

tgamblin commented on etc/spack/defaults/modules.yaml:
```diff
@@ -14,26 +14,35 @@
+      lmod:   $spack/share/spack/lmod
+    # What time of modules to use
+    enable:
+      - tcl
```
If we are changing this, should we simplify `roots:` and `enable:` to be one option?  i.e., we just enable the roots the user provided so they only specify this once?

becker33 replied:
If we do this, we will have to either change our defaults or make it slightly more difficult for user to switch between module systems.

Currently, you don't need to specify a module root to change from tcl to lmod. If we mix the two settings, then someone who merely wants to change to lmod has to choose where those modules go, rather than getting a default from Spack.

This isn't insurmountable. We could simply require it, or we could change things so you specify a single root for both types. That said, I don't think it's worth pursuing as part of this PR at the moment unless you feel strongly otherwise.

becker33 replied:
If we do this, we will have to either change our defaults or make it slightly more difficult for user to switch between module systems.

Currently, you don't need to specify a module root to change from tcl to lmod. If we mix the two settings, then someone who merely wants to change to lmod has to choose where those modules go, rather than getting a default from Spack.

This isn't insurmountable. We could simply require it, or we could change things so you specify a single root for both types. That said, I don't think it's worth pursuing as part of this PR at the moment unless you feel strongly otherwise.

tgamblin replied:
yeah let's keep it as is for now

tgamblin commented on lib/spack/spack/cmd/modules/__init__.py:
```diff
@@ -25,6 +25,11 @@
+    subparser.add_argument(
+        '-n', '--name',
+        action='store', dest='module_set_name', default='default',
+        help="Named module set to use from modules configuration."
```
can this support a comma-separated list of names?

becker33 replied:
I don't know what that would mean for `spack module tcl loads`. For the others it would be feasible though.

tgamblin commented on lib/spack/spack/cmd/modules/__init__.py:
```diff
@@ -211,11 +216,13 @@ def rm(module_type, specs, args):
     module_cls = spack.modules.module_types[module_type]
-    module_exist = lambda x: os.path.exists(module_cls(x).layout.filename)
+    module_exist = lambda x: os.path.exists(
+        module_cls(x, args.module_set_name).layout.filename)
```
I'm not seeing where module_set_name is checked against existing module sets -- can you make it so we check that against configuration, raise an error, and print the existing (i.e. selectable) module set names on error?

tgamblin replied:
This still needs addressing.

tgamblin commented on lib/spack/spack/test/data/config/config.yaml:
```diff
@@ -14,6 +14,7 @@ config:
+#  module_roots:
+#    tcl: {1}
+#    lmod: {2}
+#
```
I think this should just be deleted.  That said, I think we need code that translates the old module format into the new one, and we should have a test for that to ensure that it works.

tgamblin commented on lib/spack/spack/modules/tcl.py:
```diff
@@ -20,32 +20,34 @@
 
 
-def make_context(spec):
+def make_context(spec, module_set_name):
```
I feel like `make_context`, `make_layout`, and `make_configuration` could be consolidated.  Can the context and the layout not go in the configuration?

becker33 replied:
In terms of logical possibility, yes. Code-structure-wise, the context is an object that owns the layout and the configuration. I think reducing the complexity of this code is a separate PR if we want to take that on.

tgamblin reviewed (changes requested):
LGTM mostly. Assuming we all agree on the design here (I do) I think we now need somethign that will convert the old format to the new format, so users do not have to do that.  In addition, there should be docs for this.

But overall, I think this is a really good change and makes modules more consistent and unified across the config files (and across scopes).

There are a few places below where I wonder whether we could refactor some of the existing module logic to consolidate things -- those are probably second priority to the concerns above.

becker33 committed:
change prefix inspections for spack load to match modules

becker33 committed:
backwards compatibility for module config

becker33 committed:
clean up test config

becker33 committed:
tests for backwards compatibility of module config

becker33 committed:
change test data to match new config structure

becker33 committed:
implement post_env_write hooks for modulefile generation

new hook type: post_env_write hooks
post_env_write hooks are only executed by env.write when regenerate is True

modulefile generation for environments is a post_env_write hook
modulefile generation post-install hook skips if env present
modulefile generation in env can now be view-dependent, view is generated first

becker33 committed:
test for env with view-specific modules

becker33 committed:
add module_set_name argument to spack env loads

becker33 committed:
ensure modules only written when newly installed

becker33 committed:
update completion

becker33 committed:
flake

becker33 commented:
@tgamblin this is ready for review again

becker33 committed:
add missing test config file

becker33 committed:
make environments available in install subprocess

becker33 committed:
fix import cycle

tgamblin commented on etc/spack/defaults/modules.yaml:
```diff
@@ -34,6 +33,17 @@ modules:
+    roots:
+      tcl:    $spack/share/spack/modules
+      lmod:   $spack/share/spack/lmod
+    # What time of modules to use
```
type

tgamblin commented on lib/spack/spack/environment.py:
```diff
@@ -760,15 +764,18 @@ def _set_user_specs_from_lockfile(self):
         }
 
-    def clear(self):
+    def clear(self, re_read=False):
```
docstring

tgamblin commented on lib/spack/spack/schema/modules.py:
```diff
@@ -112,74 +116,101 @@
-# Properties for inclusion into other schemas (requires definitions)
-properties = {
-    'modules': {
+top_level_properties = {
```
Can we call this `module_config_properties`?  They're (specifically due to this PR) *not* just top-level.

A note here would also be helpful -- e.g.:

```python
#: The "real" module properties -- the actual configuration parameters.
#: They are separate from ``properties`` because they can appear both 
#: at the top level of a Spack ``modules:`` config (old, deprecated format),
#: and within a named module set (new format with multiple module sets).
```

Note that `#:` is how you do a docstring comment for variables -- sphinx will pick it up.

tgamblin reviewed (changes requested):
partial review

tgamblin reviewed (changes requested):
Another partial review

tgamblin commented on lib/spack/spack/hooks/module_file_generation.py:
```diff
@@ -11,24 +11,36 @@
 
 def post_install(spec):
-    _for_each_enabled(spec, 'write')
+    # If the spec is installed through an environment, we delegate to
```
I would rewrite this as "we skip the post_install hook and let the post_env_write hook handle this so that..."

Delegation implies that you call the post_env_write hook here, but you're really relying on the framework to call it.

tgamblin commented on lib/spack/spack/hooks/module_file_generation.py:
```diff
@@ -11,24 +11,36 @@
+    # env views and modules
+    import spack.environment  # break import cycle
+    env = spack.environment.get_env({}, '')
+    if not env:
+        _for_each_enabled(spec, 'write')
```
Nitpick:

```suggestion
    if env:
        return  # skip and leave this to post_env_write
        
    _for_each_enabled(spec, 'write')
```

The reason I'd do it this way is b/c I tend to put the exceptions first and the meat of the function last, as I expect the latter to evolve. I don't think the `return` is likely to change, but the rest of the function seems more likely to, so I'd rather it not be nested.

tgamblin commented on lib/spack/spack/environment.py:
```diff
@@ -760,15 +764,18 @@ def _set_user_specs_from_lockfile(self):
+        if not re_read:
+            # things that cannot be recreated from file
+            self.new_specs = []               # write packages for these on write()
+            self.new_installs = []            # write modules for these on write()
```
It's weird to me that we only write modules for brand new installs.  I think this makes environments nondeterministic -- people will be irritated that their modules are not regenerated.

Specifically, if I:
1. Install a bunch of stuff, or have, say, an upstream with a bunch of preexisting installations
2. Make an env with its own `env-modules:` module set and some custom settings
3. `spack install` the env

I would expect that the `env-modules` set would be generated.  It looks like it's not -- which means that we'll have weird reproducibility issues with env depending on what is already installed in a spack instance.

Is this an accurate description of the current behavior?

becker33 replied:
Yes, that's an accurate description of the current behavior, and it's not changed in this PR.

tgamblin replied:
Ok for posterity, we decided to deal with this in a future PR. The policy is its own debate and we probably need to open that up for separate comment.  We'll keep the current rewrite policy right now, but it will need to change soon b/c it doesn't make a lot of sense with multiple module sets.

tgamblin commented on lib/spack/spack/hooks/module_file_generation.py:
```diff
@@ -11,24 +11,36 @@
+
+
+def post_env_write(env):
+    for spec in env.new_installs:
```
see comment above about this.

tgamblin commented on lib/spack/spack/test/data/modules/tcl/invalid_token_in_env_var_name.yaml:
```diff
@@ -3,7 +3,7 @@ enable:
   all:
     filter:
-      environment_blacklist':
+      environment_blacklist:
```
This reminds me that Hai Ah asked us to refactor these -- we need to get rid of whitelists and blacklists in spack for inclusive naming -- that can be in a separate PR.

tgamblin reviewed (changes requested):
Ok I went through this -- only one additional large question.

becker33 committed:
check module set names for validity on module commands

becker33 committed:
typo in config comment

becker33 committed:
Environment.clear: add docstring

becker33 committed:
refactor naming of module_config_properties

becker33 committed:
module hooks: refactor post_install to have early return

becker33 committed:
env loads: default module set

tgamblin reviewed (approved):
@becker33 looks good to me.  I'm going to merge this now that the tests are passing, but you should do doc updates in a follow-on PR.

adamjstewart commented:
@becker33 @tgamblin this PR breaks `spack load` for me.

### Before

```console
$ git checkout cefbe48c89209dc3df654795644973b1885cdea4~1
$ spack load --first --sh python
# modifies the following variables:
SPACK_LOADED_HASHES
CPATH
MANPATH
PKG_CONFIG_PATH
PATH
ACLOCAL_PATH
DYLD_FALLBACK_LIBRARY_PATH
CMAKE_PREFIX_PATH
```

### After

```console
$ git checkout cefbe48c89209dc3df654795644973b1885cdea4
$ spack load --first --sh python
# modifies the following environment variables:
DYLD_FALLBACK_LIBRARY_PATH
CPATH
SPACK_LOADED_HASHES
```

My `modules.yaml` looks like:
```yaml
modules:
  enable:: []
```
but even without this file I can reproduce the bug. Any idea what could be causing this?

haampie commented:
This PR needs some documentation
